### ISSUE TITLE
Stack Overflow in Main

### ISSUE THREAD

**Issue by brandonrhan** [nus-cs2030/2122-s1#253]
<img width="520" alt="截屏2021-11-15 下午6 10 17" src="https://user-images.githubusercontent.com/92209052/141763253-d2220b07-bcd3-47b1-bd7a-6ea13d729241.png">

Does anyone else face stackoverflow when we run Main?

**Comment by yeehaoo** [comment]
Yup I have the same issue. I narrowed it down to something being wrong with my seed.get(), but I'm not sure what's going on with that. Also noticed it only happens on the 5th iteration, when n is 10^6

**Comment by brandonrhan** [comment]
I change my attribute from supplier to int and problem solved, while it means I should not write any supplier at all

**Comment by yeehaoo** [comment]
Wa but that's quite a major change... I think we'll be penalised quite heavily?

**Comment by vvidday** [comment]
Same issue here, was due to supplier as well

**Comment by huzaifa1712** [comment]
Yep, had the same issue. It's because in your next method when you chain the supplier each time with something like reassigning to () -> this.supplier.get() the number of recursive calls keep increasing. I got around it by just changing the next method to unwrap the value inside and pass a new supplier, something like

`int s = seed.get(); 
 return new Rand(() -> s, ...)
`

Though depending how you wrote your methods maybe just changing next won't be enough, not too sure.

**Comment by brandonrhan** [comment]
> Yep, had the same issue. It's because in your next method when you chain the supplier each time with something like reassigning to () -> this.supplier.get() the number of recursive calls keep increasing. I got around it by just changing the next method to unwrap the value inside and pass a new supplier, something like
> 
> `int s = seed.get(); return new Rand(() -> s, ...) `
> 
> Though depending how you wrote your methods maybe just changing next won't be enough, not too sure.

thanks a lot bro !!! I will try to fix it

**Comment by eugenelmj** [comment]
> Yep, had the same issue. It's because in your next method when you chain the supplier each time with something like reassigning to () -> this.supplier.get() the number of recursive calls keep increasing. I got around it by just changing the next method to unwrap the value inside and pass a new supplier, something like
> 
> `int s = seed.get(); return new Rand(() -> s, ...) `
> 
> Though depending how you wrote your methods maybe just changing next won't be enough, not too sure.

I was facing the exact same issue here and I got stuck for quite a while; but this solution here of seed.get() did help to resolve it for me (and the difference was quite drastic in terms of reducing the overflow) so kudos to Huzaifa for this! 

For anyone else who uses Supplier in their implementation and is facing the stack overflow issue, do try this solution and it should help.
